flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java

```diff
@@ -123,9 +123,18 @@ private void handleDeploymentFailure(ExecutionVertex executionVertex, JobExcepti
     }
 
     @Override
-    public void notifyNewResourcesAvailable() {
-        if (context.canScaleUp(getExecutionGraph())) {
-            getLogger().info("New resources are available. Restarting job to scale up.");
+    public void onNewResourcesAvailable() {
+        maybeRescale();
+    }
+
+    @Override
+    public void onNewResourceRequirements() {
+        maybeRescale();
+    }
```

Comment:
1. Should there be a timeout to eventually satisfy the requirements? (similar to one in `WaitingForResources`)
I guess in `onNewResourcesAvailable()`, new resources can be ignored indefinitely, but we can't afford that for requirements?

2. Is `maybeRescale()` indeed suitable for new requirements?
In case when some vertices were up-scaled and some downscaled, the cumulative parallelism won't change (that's what is eventually checked in `AdaptiveScheduler.shouldRescale` from `maybeRescale`).
Or am I missing something?

Reply:
>  Or am I missing something?

I don't think you are; this is a good catch.

Reply:
> In case when some vertices were up-scaled and some downscaled, the cumulative parallelism won't change 

This now raises an interesting question w.r.t. `minParallelismIncrease`.

If the increase is smaller then the defined minimum, should we rescale? In https://github.com/apache/flink/pull/22169#discussion_r1144769916 we said that a scale down should happen regardless of the minimum.
So what takes precedence now? :thinking: 

Maybe we can stick to "scale down whenever possible" and revisit this separately. We need to think through the configuration to handle this, and I don't want to block the PR. It should ultimately end up being a small change contained to the `RescalingController`.
I do think it's likely that someone will run into this; but even in the current state you could work around it by separating scale up/down operations.

Reply:
> Should there be a timeout to eventually satisfy the requirements?

Yes. I'd revisit this along with `minParallelismIncrease` because it's all about configuring the scaling behavior in certain edge-cases.

Reply:
I think there's no contradiction with "scale down whenever possible" if we interpret it as "scale down whenever **cumulative** parallelism decreases".

It should be fine to address this later. Shall we create a ticket for that? (maybe one catch-all ticket for now)

Reply:
Or do you mean it contradicts the current behavior of Reactive Scheduling?

I actually think it doesn't. With RS, there are only changes in resources, which ultimately result in changing cumulative parallelism. So there will be no change in RS behavior.

Reply:
> do you mean it contradicts the current behavior of Reactive Scheduling?

Reactive mode isn't affected because there the actual requirements don't change.

> I think there's no contradiction with "scale down whenever possible" if we interpret it as "scale down whenever cumulative parallelism decreases".

That would indeed not be contradictory, but not really a desirable behavior? Not without some fallback condition at least... (like a timeout).

> Shall we create a ticket for that? (maybe one catch-all ticket for now)

Will do.

Reply:
https://issues.apache.org/jira/browse/FLINK-31608

Reply:
Thanks!
Should we also create a ticket for 
> 1. Should there be a timeout to eventually satisfy the requirements? (similar to one in WaitingForResources)
I guess in onNewResourcesAvailable(), new resources can be ignored indefinitely, but we can't afford that for requirements?

?

Reply:
I think that can be part of FLINK-31608. If we extend the logic that limits rescaling we should also discuss such a fallback.
I'll add a separate bullet point for that.

Reply:
Thanks!